cypress/integration/DropdownMenu.spec.ts

```diff
@@ -0,0 +1,202 @@
+        movePointerTo('Inbox');
+        movePointerTo('Bookmarks →');
+        cy.findByText('Inbox').should('exist');
+      });
+
+      it('should close when pointer moves to an item or sibling trigger in parent menu', () => {
+        movePointerTo('Bookmarks →');
+        movePointerTo('Inbox');
+        movePointerTo('New Window');
+        cy.findByText('Inbox').should('not.exist');
+
+        movePointerTo('Bookmarks →');
+        movePointerTo('Inbox');
+        movePointerTo('Tools →');
+        cy.findByText('Inbox').should('not.exist');
+      });
+
+      it('should remain open when clicking trigger but close when clicking item', () => {
+        movePointerTo('Bookmarks →').click();
+        cy.findByText('Inbox').click();
+        cy.findByText('Inbox').should('not.exist');
+      });
+
+      it('should not open when trigger is disabled', () => {
+        movePointerTo('History →');
+        cy.findByText('Github').should('not.exist');
+
+        cy.findByText('History →').trigger('keydown', { key: ' ' });
+        cy.findByText('Github').should('not.exist');
+
+        cy.findByText('History →').trigger('keydown', { key: 'Enter' });
+        cy.findByText('Github').should('not.exist');
+      });
+
+      it('should open on space and enter key press', () => {
+        cy.findByText('Bookmarks →').trigger('keydown', { key: ' ' });
+        cy.findByText('Inbox').should('exist');
+
+        movePointerTo('New Tab');
+
+        cy.findByText('Bookmarks →').trigger('keydown', { key: 'Enter' });
+        cy.findByText('Inbox').should('exist');
+      });
+
+      it('should close on escape key press', () => {
+        movePointerTo('Bookmarks →');
+        movePointerTo('Inbox').type('{esc}');
+        cy.findByText('Inbox').should('not.exist');
+
+        cy.findByText('Open').click();
+        movePointerTo('Bookmarks →');
+        movePointerTo('Modulz →').type('{esc}');
+        cy.findByText('Inbox').should('not.exist');
+      });
+
+      it('should not focus first item when opening via pointer', () => {
+        movePointerTo('Bookmarks →').then(($trigger) => {
+          const associatedMenuId = $trigger.attr('aria-controls');
+          cy.get(`[id="${associatedMenuId}"]`).children().first().should('not.be.focused');
+        });
+      });
+    });
+
+    describe('in LTR configuration', () => {
+      it('should remain open when exiting trigger towards submenu', () => {
+        cy.findByText('Open').click();
+
+        movePointerTo('Bookmarks →');
+        cy.findByText('Inbox').should('exist');
+
+        exitPointerLeftToRight('Bookmarks →');
+        cy.findByText('Inbox').should('exist');
+      });
+
+      it('should close when exiting trigger away from submenu', () => {
+        cy.findByText('Open').click();
+
+        movePointerTo('Bookmarks →');
+        cy.findByText('Inbox').should('exist');
+
+        exitPointerRightToLeft('Bookmarks →');
+        cy.findByText('Inbox').should('not.exist');
+      });
```

Comment:
We tend to write our tests like "given, when, then" (nested describes) which really helps avoid test setup duplication, and with one assertion per test to isolate failures (making them easier to debug/fix): 

```jsx
describe('given an open menu', () => {
	beforeEach(() => {
		cy.findByText('Open').click();
	});

	describe('when moving mouse pointer on submenu trigger', () => {
		beforeEach(() => {
			movePointerTo('Bookmarks →');
		});

		it('should open the submenu', () => {
			cy.findByText('Inbox').should('exist');
		});

		it('should not focus first item', () => {
        	cy.findByText('Inbox').should('not.be.focused');
      	});

		describe('and moving pointer towards submenu', () => {
			beforeEach(() => {
				exitPointerLeftToRight('Bookmarks →');
			});

			it('should keep the submenu open', () => {
				cy.findByText('Inbox').should('exist');
			});

			it('should keep the main menu open', () => {
				cy.findByText('Bookmarks →').should('exist');
			});
		});

		describe('and moving pointer away from submenu', () => {
			beforeEach(() => {
				exitPointerRightToLeft('Bookmarks →');
			});

			it('should close the submenu', () => {
				cy.findByText('Inbox').should('not.exist');
			});

			it('should close the main menu', () => {
				cy.findByText('Bookmarks →').should('exist');
			});
		});

		describe('and moving pointer to trigger sibling item', () => {
			beforeEach(() => {
				movePointerTo('New Window');
			});

			it('should close the submenu', () => {
				cy.findByText('Inbox').should('not.exist');
			});

			it('should keep the main menu open', () => {
				cy.findByText('Bookmarks →').should('exist');
			});
		});

		describe('and moving pointer to trigger sibling trigger', () => {
			beforeEach(() => {
				movePointerTo('Tools →');
			});

			it('should close the submenu', () => {
				cy.findByText('Inbox').should('not.exist');
			});

			it('should keep the main menu open', () => {
				cy.findByText('Bookmarks →').should('exist');
			});
		});

		describe('and clicking a submenu item', () => {
			beforeEach(() => {
	        	cy.findByText('Inbox').click();				
			});

			it('should close the submenu', () => {
				cy.findByText('Inbox').should('not.exist');
			});

			it('should close the main menu', () => {
				cy.findByText('Bookmarks →').should('not.exist');
			});
		});

		describe('and pressing escape', () => {
			beforeEach(() => {
				cy.get('body').type('{esc}');
			});

			it('should close the submenu', () => {
				cy.findByText('Inbox').should('not.exist');
			});

			it('should close the main menu', () => {
				cy.findByText('Bookmarks →').should('not.exist');
			});
		});
	});

	describe('when in RTL mode', () => {
		beforeEach(() => {
			cy.findByText('Right-to-left').click();
		});

		describe('and moving pointer ....', () => {});
	});
});
```

Hopefully there are enough examples here to explain what I mean but happy to have a call to run through it if it's not clear 🙂

Reply:
Thanks for this! your example has already highlighted some omissions that would help the tests.

On the "given, when, then" point, I really like this! but there are some caveats with cypress which can make it challenging. The [best practice doc](https://docs.cypress.io/guides/references/best-practices#Creating-tiny-tests-with-a-single-assertion ) sums some of those up.

The other issue is the way that assertions in Cypress have a concept of [retry-ability](https://docs.cypress.io/guides/core-concepts/retry-ability), it’s kinda their answer to common flake in e2e, which for the most part makes tests more reliable but using it with this suggested approach can make failures sometimes less clear.

If you’re free today would be great to jump on a call to discuss, perhaps we can come out with an improved structure while working around the constraints 🙂

Reply:
Ah, thanks for the link. Seems they clearly recommend against this but that makes my heart sad because I find this sooooo much easier to reason about. Also, why would they even provide these APIs if it's an anti-pattern 🤔 

Their reasons against it seem more opinionated except for the last point about perf. Obviously that's an issue but how much slower are we talking?

With the current tests it's not clear which part of the test is the bit described by the test and which parts are just repeated set up 🙈 

Let's have a catch up on this today so I can maybe understand how much of an issue the issues they describe really are.

Reply:
We caught up on this and we're going to stick with the general approach @andy-hook has here but try make the code a bit more self documented to help with some of my confusions. 

Andy also helped me realise that it's not hugely important to be able to reason about the code because with Cypress we can step through the test visually which is super helpful.